### ISSUE TITLE
Enhancement: return-await "always" should require await on "any" type

### ISSUE THREAD

**Issue by matthieusieben** [typescript-eslint/typescript-eslint#5117]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/return-await/

### Description

I propose the ability to require to use `return await` on values with the `any` type in async functions.

### Fail

```typescript
const foo = (): any => 0
const bar = async () => {
  try {
    return foo() // Because `any` can be a promise, it should be awaited
  } catch {
    return 42
  }
}
```

### Pass

```typescript
const foo = (): any => 0

const bar = async () => {
  try {
    return await foo()
  } catch {
    return 42
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
I can't repro the rule reporting any error. Could you please post a link to a [playground reproduction](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=MYewdgzgLgBAZiEMC8MAUBKAXDAhmATxQD4YAGAWAChRJYAjXAJxTwgLGHQxJgG9qMGFCZEBVIUKYBTKAFcmYeIkwwA9GpgAhacFxyI0mAAN8BYzD1L6R3DAAOTEAFsAloYA0MV7AgALEDkAGwATGBs8AHdcH2kQwRgAX0tcKGA-fgSpWQUlABYAJgTE6hKqIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRfWaJ2gQwDufSvnRRE0aAHtokcGAC+IBUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) that does?

In theory the rule should already work like this 🤔.

**Comment by JoshuaKGoldberg** [comment]
🤔 hmm. Other than warning away from `any` types with the `@typescript-eslint/no-unsafe-*` rules, we really generally try to stay away from `any`s. What's the context for why you want this?

**Comment by matthieusieben** [comment]
The main reason I get this issue is because I use ts-eslint across a big codebase (monorepo) that includes code in vue, JS, etc.

I want to enforce `return await` whenever a promise is used but some time the typing system is not always able to catch that up.

Since `return await` of non promise does not really cause runtime problems, I think that there should be a way to make `return await any` be "valid"

**Comment by bradzacher** [comment]
> but some time the typing system is not always able to catch that up

What do you mean by this, exactly?
Do you mean people are specifying bad types? Or do you mean TS is broken?

> Since return await of non promise does not really cause runtime problems

As mentioned on your other issue #5118, it isn't as harmless as it might seem.


---

As Josh mentioned - we have lint rules to help warn against accidental or intentional anys in the codebase - `no-explicit-any` and the `no-unsafe-*` group of rules.
Those would be a much better route of enforcing safety in the codebase, rather than just allowing anys and patching the holes it creates down the line.

Similar to #5118 - I'm not sure I see the value in allowing this case in the rule.

**Comment by matthieusieben** [comment]
When linting a Vue file which uses methods from mixing, TS does not "know" about the signature of those methods. In that case, TS just uses "any" as return type, which yields an "return-await" eslnt error.

**Comment by matthieusieben** [comment]
The feat here is to allow a function that does return a promise, but is incorrectly detected as returning `any`, tu be awaited on (eg. When used in a try catch)

The following code should not yield an error:
```js
// This is not properly typed due to limitations in TS (e.g. defined through a Vue "mixin".
const foo = (): any => Promise.resolve(0)

const bar = async () => {
  try {
    return await foo()
  } catch {
    return 42
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
> This is not properly typed due to limitations in TS

Therein lies the rub: we have to go with whatever TypeScript gives us.

I'm going to go ahead and close this issue as we don't want to be in the business of assuming the worst or trying to seriously work with `any` types.

Summarizing advice from the thread:
* In general, we recommend using the `@typescript-eslint/no-unsafe-*` rules to avoid dangerous `any` code
* If there's an issue preventing proper types being inferred, such as Vue parsing shenanigans, let's try to resolve that issue specifically
